File: 42-trapping-rain-water/42-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        
        int left[] = new int[n];
        int right[] = new int[n];
        
        int maxLeft = height[0];
        for(int i=0;i<n;i++){
            maxLeft = Math.max(maxLeft,height[i]);
            left[i] = maxLeft;
        }
        
        int maxRight = height[n-1];
        for(int i=n-1;i>=0;i--){
            maxRight = Math.max(maxRight,height[i]);
            right[i] = maxRight;
        }
        
        int ans=0;
        for(int i=0;i<n;i++){
            ans += Math.min(left[i],right[i]) - height[i];
        }
        
        return ans;
    }
}

File: 42-trapping-rain-water/42-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int num = height.size();
        if(num==0)
            return 0;
        vector<int> leftMax(num),rightMax(num);
        int temp = height[0];
        for(int i=0;i<num;i++){
            if(height[i]>temp)
                temp=height[i];
            leftMax[i]=temp;
        }
        temp = height[num-1];
        for(int i=num-1;i>=0;i--){
            if(height[i]>temp)
                temp=height[i];
            rightMax[i]=temp;
        }
        int sum=0;
        for(int i=0;i<num;i++)
            sum+=min(leftMax[i],rightMax[i]) - height[i];
        return sum; 
    }
};